project1/thread_pool/thread_pool_interface.h

```diff
@@ -16,26 +17,30 @@ class IThreadPool {
 
   /**
    * @brief Adds a new task to the pool.
-   * @param task The task to add. The pool will take ownership of it.
-   * @return A handle to the task in the pool.
+   * @param task The task to add.
    */
-  virtual Task::Handle AddTask(std::unique_ptr<Task>* task) = 0;
+  virtual void AddTask(const std::shared_ptr<Task>& task) = 0;
 
   /**
    * @brief Gets the current status of task.
-   * @param task_handle The handle of the task.
+   * @param task The task..
    * @return The status of the task.
    */
-  virtual Task::Status GetTaskStatus(const Task::Handle& task_handle) = 0;
+  virtual Task::Status GetTaskStatus(const std::shared_ptr<Task>& task) = 0;
```

Comment:
Why the change here? Just curious.

Reply:
In my initial implementation, I was using `std::unique_ptr` for tasks and I actually had the pool take complete ownership of the task. This meant that the main thread had no access to the task after calling `AddTask`. Because of that, all the methods like `GetTaskStatus` had to use the task handle, which could easily be saved, as opposed to the task itself.

However, as I actually tried to use this implementation, I quickly realized that it was annoying. There are, in fact, legitimate reasons why you want the main thread to still have access to submitted tasks. For instance, if you have tasks that produce results, the "correct" way to implement that is to store the result as a member of the `Task` itself and then query the result in the main thread once the task is complete.

Therefore, I transitioned to using `std::shared_ptr` for tasks instead, which allows for both paradigms: If you don't want to keep the task around in the main thread, you don't have to, and the thread pool will automatically take sole ownership of the task, just as before. However, if you want to keep it around, you can do that too, and now the main thread and pool will share ownership.

Once it became possible to hold onto `Task` objects in the main thread, using the task handle for everything became unnecessary, and the API was much more unwieldy than just passing the actual `Task` object. That's why I ultimately changed it.